src/anyp/Uri.cc

```diff
@@ -81,6 +81,48 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    const auto end = buf.end();
+    for (auto i = buf.begin(); i != end; ++i) {
+        const auto ch = *i;
+        if (ch != '%') {
+            output.append(ch);
+            continue;
+        }
+        // ch is '%', let's look ahead
+        ++i;
+        if (i == end)
+            throw TextException("incomplete %-encoded triplet at end of input", Here());
+        const auto ch1 = *i;
+        ++i;
+        if (i == end)
+            throw TextException("incomplete %-encoded triplet at end of input", Here());
+        const auto ch2 = *i;
+        const auto hex1 = HextDigitToInt(ch1);
+        const auto hex2 = HextDigitToInt(ch2);
+        if (hex1 == -1 || hex2 == -1)
+            throw TextException("invalid %-encoded triplet", Here());
+
+        const auto decoded = hex1 << 4 | hex2;
+        output.append(decoded);
+    }
```

Comment:
This being a parser, we should use a `Tokenizer`. 
The logic of this loop then becomes:
```suggestion
    Parser::Tokenizer tok(buf);
    while(!tok.atEnd()) {
        static const auto unencodedOctets = CharacterSet("unencoded","%").compliment();
        output.append(tok.prefix("unencoded", unencodedOctets));

        if (tok.atEnd())
            continue;

        if (tok.skipOne('%')) {
            int64_t hex1 = 0, hex2 = 0;
            if (tok.int64(hex1, 16, false, 1) && tok.int64(hex2, 16, false, 1)) {
                output.append(char((hex1<<4)|hex2));
                continue;
            }
        }

        throw TextException("invalid %-encoded triplet", Here());
    }
```

Reply:
Switching to Tokenizer is a step forward and will reduce the amount of work I will need to do in this PR. Keeping the following rules of thumb in mind may reduce that work further:

* avoid atEnd() checks inside the parsing loop
* avoid `if` statements inside the parsing loop[^1]
* avoid Tokenizer::int64() when parsing strict grammars
* use names of grammar elements from the RFC
* keep in mind that Tokenizer::foo(description, ...) methods throw on bad and insufficient input
* InsufficientInput exceptions must not spread beyond parsing code[^2]

[^1]: Some `if` statements are unavoidable, of course. The proposed loop has three (including one complex statement that should really be counted as two). I hope we can reduce that count to one.

[^2]: Unlike TextExceptions, InsufficientInput exceptions are dedicated to parsing needs. They are not general-purpose exceptions. If "final" parsing code decides to use helper parser classes/functions that may throw InsufficientInput exceptions, that parsing code is responsible for shielding the rest of Squid from them.

Reply:
done, using a slightly different flow than suggested by @yadij